File: tasks/saliency/models/layers/slice_pool_layer/src/slice_pool_layer.c

#include <TH/TH.h>
#include <math.h>
#include <stdio.h>


//-------- Max Pooling Functions

// Max Pooling Forward function
int slice_pool_max_forward(THFloatTensor *data_tensor, THIntTensor *slice_idx_tensor, int num_slice, int num_batch, int channels, int num_points, THFloatTensor *output_tensor, THIntTensor *pool_mask_tensor)
{
    // data:                    num_batch, channels,  num_points  , 1
    // output:                  num_batch, channels,  num_slice, 1
    // slice_idx:               num_batch, num_points
    
    float * data = THFloatTensor_data(data_tensor);
    int * slice_idx = THIntTensor_data(slice_idx_tensor);
    float * output = THFloatTensor_data(output_tensor);
    int * pool_mask = THIntTensor_data(pool_mask_tensor);
    
    // Max Pooling
    int n, c, m;
    for (n = 0; n < num_batch; n++) {
        for (c = 0; c < channels; c++) {
            for (m = 0; m < num_points; m++) {
                
                // get slice index
                int idx = slice_idx[n * num_points + m];   // slice_idx[n, m]
            
                // get output index
                int output_idx =  n * channels * num_slice + c * num_slice + idx; // output[n, c, idx, 0]

                // get pool mask index
                int pool_mask_idx =  n * channels * num_slice + c * num_slice + idx; // output[n, c, idx, 0]
                
                // get input index
                int input_index =  n * channels * num_points + c * num_points + m; // data[n, c, m, 0]
                
                if (data[input_index] > output[ output_idx ])
                {
                    output[output_idx] = data[input_index];
                    pool_mask[pool_mask_idx] = input_index;
                }
                
            }
        }
    }
    return 1;
}




// Max Pooling Backward function
int slice_pool_max_backward(THFloatTensor *top_grad_tensor, THIntTensor *pool_mask_tensor, int num_slice, int num_batch, int channels, THFloatTensor *output_tensor )
{
    // top grad:    num_batch, channels, slice, 1
    // pool mask:   num_batch, channels, slice, 1
    // output:      num_batch, channels, num_points, 1
    
    float * top_grad = THFloatTensor_data(top_grad_tensor);
    int * pool_mask = THIntTensor_data(pool_mask_tensor);
    float * output = THFloatTensor_data(output_tensor);
    
    int n, c, m;
    for (n = 0; n < num_batch; n++) {
        for (c = 0; c < channels; c++) {
            for (m = 0; m < num_slice; m++) {
                
                int top_index = n * channels * num_slice + c * num_slice + m; // output[n,c,m,0]
                int bottom_index = pool_mask[top_index];
                
                if (bottom_index != -1) {
                    output[bottom_index] += top_grad[top_index];
                }
                
            }
        }
    }
    return 1;
}




//-------- Avg Pooling Functions

// Average Pooling Forward function
int slice_pool_avg_forward(THFloatTensor *data_tensor, THIntTensor *slice_idx_tensor, THIntTensor *slice_counts_tensor , int num_slice, int num_batch, int channels, int num_points, THFloatTensor *output_tensor)
{
    // data:            num_batch, channels, num_points    , 1
    // output:          num_batch, channels, slice  , 1
    // slice_idx:       num_batch, num_points
    // slice_counts:    num_batch, slice
    
    float * data = THFloatTensor_data(data_tensor);
    int * slice_idx = THIntTensor_data(slice_idx_tensor);
    int * slice_counts = THIntTensor_data(slice_counts_tensor);
    float * output = THFloatTensor_data(output_tensor);
    
    
    // Max Pooling
    int n,c,m;
    for (n = 0; n < num_batch; n++) {
        for (c = 0; c < channels; c++) {
            for (m = 0; m < num_points; m++) {
                
                // get slice idx
                int c_idx = slice_idx[ n * num_points + m ] ; // slice_idx[n, m]
                
                // get slice counts idx
                float slice_count = slice_counts[ n * num_slice + c_idx ]; // slice_counts[n, c_idx]
                
                // get output idx
                int output_idx = n * channels * num_slice + c * num_slice + c_idx; // output[n, c, c_idx]
                
                // get input idx
                int input_idx = n * channels * num_points + c * num_points + m; // data[n, c, m, 0]
                
                output[ output_idx ] += data[input_idx] / slice_count;
                
            }
        }
    }
    return 1;
}



// Avg Pooling Backward function
int slice_pool_avg_backward(THFloatTensor *top_grad_tensor, THIntTensor *slice_idx_tensor, THIntTensor *slice_counts_tensor , int num_slice, int num_batch, int channels, int num_points, THFloatTensor *output_tensor )
{
    // top grad:        num_batch, num_slice,  1
    // slice_idx:       num_batch, num_points
    // slice_counts:    num_batch, slice
    // output:          num_batch, channels, num_points, 1
    
    float * top_grad = THFloatTensor_data(top_grad_tensor);
    int * slice_idx = THIntTensor_data(slice_idx_tensor);
    int * slice_counts = THIntTensor_data(slice_counts_tensor);
    float * output = THFloatTensor_data(output_tensor);
    
    int n, c, m;
    for (n = 0; n < num_batch; n++) {
        for (c = 0; c < channels; c++) {
            for (m = 0; m < num_points; m++) {
                
                // get slice index
                int c_idx = slice_idx[ n * num_points + m ] ; // slice_idx[n, m]
                
                // get slice counts idx
                float slice_count = slice_counts[ n * num_slice + c_idx ]; // slice_counts[n, c_idx]
                
                // get top idx
                int top_idx = n * channels * num_slice + c * num_slice + c_idx; // output[n, c, c_idx]

                // get bottom idx
                int bottom_idx = n * channels * num_points + c * num_points + m; // data[n, c, m, 0]
                
                if (bottom_idx != -1) {
                    output[bottom_idx] += top_grad[top_idx] / slice_count;
                }
            }
        }
    }
    return 1;
}

File: tasks/saliency/models/layers/slice_pool_layer/slice_pool_layer.py
import torch
from torch.autograd import Function
import sys
import os
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
from ext_pool import slice_pool_layer
from torch.autograd import Variable


class Slice_Pool(Function):
    def __init__(self, num_slice):
        super().__init__()
        self.num_slice = num_slice
    
    def forward(self, input, slice_idx_mat, pool_type, slice_counts):
        
        assert input.is_contiguous()
        
        if pool_type.numpy()[0] == 1:
            self.pool_type = "Max_Pool"
        elif pool_type.numpy()[0] == 2:
            self.pool_type = "Avg_Pool"
            
        num_batch, channels, num_points, _ = input.size()
        num_slice = self.num_slice
        if num_slice == None:
            num_slice = int(slice_idx_mat.max()) + 1
            
        self.num_points = num_points
        
        if self.pool_type == "Max_Pool":
            out = torch.ones(num_batch, channels, num_slice, 1) * 0
        else:
            out = torch.zeros(num_batch, channels, num_slice, 1)
    
        pool_mask = torch.IntTensor(num_batch, channels, num_slice, 1)
        pool_mask[:,:,:,:] = -1

        # Max_Pool
        if self.pool_type == "Max_Pool":
            if not input.is_cuda:
                slice_pool_layer.slice_pool_max_forward(input, slice_idx_mat, num_slice, num_batch, channels, num_points, out, pool_mask)
                self.pool_mask = pool_mask
            else:
                out = out.cuda()
                pool_mask = pool_mask.cuda()
                slice_pool_layer.slice_pool_max_forward_cuda(input, slice_idx_mat, num_slice, num_batch, channels, num_points, out, pool_mask)

            self.pool_mask = pool_mask


        # Avg_Pool
        elif self.pool_type == "Avg_Pool":
            self.slice_idx_mat = slice_idx_mat
            self.slice_counts = slice_counts
            if not input.is_cuda:
                slice_pool_layer.slice_pool_avg_forward(input, slice_idx_mat, slice_counts, num_slice, num_batch, channels, num_points, out)
            else:
                out = out.cuda()
                slice_pool_layer.slice_pool_avg_forward_cuda(input, slice_idx_mat, slice_counts, num_slice, num_batch, channels, num_points, out)
            
        
        return out


    def backward(self, grad_top):
        num_batch, channels, num_slice, _ = grad_top.size()
        grad_bottum = torch.zeros(num_batch, channels, self.num_points, 1).cuda()
        # Max Pooling
        if self.pool_type == "Max_Pool":
            if not grad_top.is_cuda:
                slice_pool_layer.slice_pool_max_backward(grad_top, self.pool_mask, num_slice, num_batch, channels, grad_bottum)
            else:
                # grad_bottum = grad_bottum.cuda()
                slice_pool_layer.slice_pool_max_backward_cuda(grad_top, self.pool_mask, num_slice, num_batch, channels, grad_bottum)


    
        # Avg_Pool
        elif self.pool_type == "Avg_Pool":
            if not grad_top.is_cuda:
                slice_pool_layer.slice_pool_avg_backward(grad_top, self.slice_idx_mat, self.slice_counts, num_slice, num_batch, channels, self.num_points, grad_bottum)
            else:
                grad_bottum = grad_bottum.cuda()
                slice_pool_layer.slice_pool_avg_backward_cuda(grad_top, self.slice_idx_mat, self.slice_counts, num_slice, num_batch, channels, self.num_points, grad_bottum)
                
            
        return grad_bottum, None, None, None





class SP(torch.nn.Module):
    def __init__(self, pool_type, num_slice=None):
        super(SP, self).__init__()
        if pool_type == "Max_Pool":
            self.pool_type = Variable(torch.ones(1,1)) * 1
        elif pool_type == "Avg_Pool":
            self.pool_type = Variable(torch.ones(1,1)) * 2
        
        self.sp=Slice_Pool(num_slice)
    #
    def forward(self, input, slice_idx_mat, slice_counts=torch.autograd.Variable(None)):
        return self.sp(input, slice_idx_mat, self.pool_type, slice_counts)


